### ISSUE TITLE
Grouping fails when dragging column - Cannot read property 'types' of undefined (undefined = dataTransfer)

### ISSUE THREAD

**Issue by mpkelly** [adazzle/react-data-grid#1014]
### Which version of React JS are you using?

✅ Officially supported ✅
- [ ] v15.4.x

⚠️ Not officially supported, expect warnings ⚠️
- [ ] v15.5.x
- [x] v15.6.x

☣️ Not officially supported, expect warnings and errors ☣️
- [ ] v16.x.x

---

### Which browser are you using?

✅ Officially supported ✅
- [ ] IE 9 / IE 10 / IE 11
- [ ] Edge
- [x] Chrome

⚠️ Not officially supported, but "should work" ⚠️
- [ ] Firefox
- [ ] Safari

---

### I'm submitting a ...

- [x] 🐛 Bug Report
- [ ] 💡 Feature Request

> 👋 Need general support? Not sure about how to use React itself, or how to get started with the Grid?
> Please do not submit support request here. Instead see

> https://github.com/adazzle/react-data-grid/blob/master/CONTRIBUTING.md

---

### Issue Details

I would like to support sorting, filtering and grouping and have created the implementation below that I based on the official examples. However, when I try to drag a column to I get the following error:

```
bundle.js:176335 Uncaught TypeError: Cannot read property 'types' of undefined
    at matchNativeItemType (bundle.js:176335)
    at HTML5Backend.handleTopDragStart (bundle.js:175810)
    at GestureDetector.Instance.triggerEvent [as trigger] (bundle.js:80425)
    at Object.dragGesture [as handler] (bundle.js:80580)
    at Object.triggerGesture (bundle.js:80128)
    at Array.forEach (<anonymous>)
    at Object.each (bundle.js:79389)
    at Object.detect (bundle.js:80125)
    at Object.doDetect (bundle.js:79848)
    at HTMLDocument.onTouchHandler (bundle.js:79784)
```

The function that errors is:

```JavaScript
	function matchNativeItemType(dataTransfer) {
		var dataTransferTypes = Array.prototype.slice.call(dataTransfer.types || []); //ERROR

		return Object.keys(nativeTypesConfig).filter(function (nativeItemType) {
			var matchesTypes = nativeTypesConfig[nativeItemType].matchesTypes;

			return matchesTypes.some(function (t) {
				return dataTransferTypes.indexOf(t) > -1;
			});
		})[0] || null;
	}

```

My code

```Typescript
import * as React from "react";
import * as ReactDataGrid from "react-data-grid";
import {ToolsPanel, Data, Draggable} from "react-data-grid-addons";

export interface DashboardTaskGridProps {

}

export interface DashboardTaskGridState {
  rows:any[];
  filters: any;
  sortColumn:string;
  groupBy:any[];
  expandedRows:any;
  sortDirection:string;
}

export class DashboardTaskGrid extends React.Component<DashboardTaskGridProps, DashboardTaskGridState> {

  private columns: any[];
  public grid: any;

  constructor(props: DashboardTaskGridProps) {
    super(props);
    this.columns = [
      { key: 'id', name: 'ID', draggable: true},
      { key: 'title', name: 'Title', sortable: true, filterable:true, draggable: true},
      { key: 'count', name: 'Count', sortable: true, filterable:true, draggable: true} ];

    this.state = {
      rows: this.createRows(),
      filters: {},
      groupBy: [],
      sortColumn:null,
      sortDirection: null,
      expandedRows: {}
    };
  }

  createRows = () => {
    let rows = [];
    for (let i = 1; i < 1000; i++) {
      rows.push({
        id: i,
        title: 'Title ' + i,
        count: i * 1000
      });
    }

   return rows;
  };

  handleGridSort = (sortColumn, sortDirection) => {
    this.setState({ sortColumn: sortColumn, sortDirection: sortDirection });
  };

  getSize = () => {
    return this.getRows().length;
  };

  getRows = () => {
    return Data.Selectors.getRows(this.state);
  };

  rowGetter = (index) => {
    const rows = this.getRows();
    return rows[index];  };

  handleColumnFilter = (filter:any) => {
    let newFilters = {...this.state.filters};
    if (filter.filterTerm) {
      newFilters[filter.column.key] = filter;
    } else {
      delete newFilters[filter.column.key];
    }
    this.setState({ filters: newFilters });
  };

  onColumnGroupAdded = (colName) => {
    let columnGroups = this.state.groupBy.slice(0);
    let activeColumn = this.columns.find((c) => c.key === colName)
    let isNotInGroups = columnGroups.find((c) => activeColumn.key === c.name) == null;
    if (isNotInGroups) {
      columnGroups.push({key: activeColumn.key, name: activeColumn.name});
    }

    this.setState({groupBy: columnGroups});
  };

  onColumnGroupDeleted = (name) => {
    let columnGroups = this.state.groupBy.filter(function(g){
      return typeof g === 'string' ? g !== name : g.key !== name;
    });
    this.setState({groupBy: columnGroups});
  };

  onRowExpandToggle = ({ columnGroupName, name, shouldExpand }) => {
    let expandedRows = Object.assign({}, this.state.expandedRows);
    expandedRows[columnGroupName] = Object.assign({}, expandedRows[columnGroupName]);
    expandedRows[columnGroupName][name] = {isExpanded: shouldExpand};
    this.setState({expandedRows: expandedRows});
  };

  createToolbar = () => {
    return (
      <ToolsPanel.AdvancedToolbar>
        <ToolsPanel.GroupedColumnsPanel
          groupBy={this.state.groupBy}
          onColumnGroupAdded={this.onColumnGroupAdded}
          onColumnGroupDeleted={this.onColumnGroupDeleted}/>
      </ToolsPanel.AdvancedToolbar>
    );
  };

  render() {
    return (
      <Draggable.Container>
        <ReactDataGrid
          ref={ node => {
            this.grid = node;
            this.grid.onToggleFilter();
          } }
          onGridSort={this.handleGridSort}
          onAddFilter={this.handleColumnFilter}
          onRowExpandToggle={this.onRowExpandToggle}
          enableDragAndDrop={true}
          columns={this.columns}
          toolbar={this.createToolbar()}
          rowGetter={this.rowGetter}
          rowsCount={this.getSize()}
          canFilter={true}
          onClearFilters={()=>this.setState({filters: {}})}
          minHeight={500} />
        </Draggable.Container>
    )
  }
}

```
Is this a bug or am I missing a dependency somewhere?

**Comment by mpkelly** [comment]
I guess no one can see an issue with my code. I couldn't either - it is basically the example code merged into one. 

Can anyone point me to a working example that uses sorting, filtering and merging? Maybe I can try to start from that instead. 

Can someone also confirm that I do not need to load any external dependencies (css, jQuery etc) beyond those that are imported in my example code?

**Comment by wfgmiami** [comment]
Hi, I do not have an answer to your question but I actually have a question for you - you are saying that you have created the implementation based on the official examples - where did you find the official examples?
The demo link at http://adazzle.github.io/react-data-grid/ does not seem to work - is it that link where the official example are shown?

Thanks

**Comment by mpkelly** [comment]
Hi @wfgmiami, 

The examples can be found here:

http://adazzle.github.io/react-data-grid/examples.html#/basic

or, if you are just looking for the source code, then you can go to:

https://github.com/adazzle/react-data-grid/tree/master/packages/react-data-grid-examples/src/scripts

Mike

**Comment by wfgmiami** [comment]
@mpkelly Thank you for getting back. I was also trying to get all 3 functionalities  - sorting, filtering and grouping. I have the sorting and grouping working but the filtering is not showing in the toolbar now since i have the CustomToolbar that supports (looks like) only the grouping but not the filtering.
Not sure I see implementation with all 3 in the demos.

**Comment by mpkelly** [comment]
@wfgmiami I had come to same conclusion myself. I don't think these three features are compatible without modifying the project.  It would be nice if someone from the project could confirm this and document it somewhere?

**Comment by wfgmiami** [comment]
@mpkelly I get it to work - the code is 99% what is the demo code for filtering and grouping. The change I made was in the following code:
  ToolsPanel: { AdvancedToolbar: Toolbar, GroupedColumnsPanel }

Instead of that, I have:
import { ToolsPanel } from 'react-data-grid-addons';
const { GroupedColumnsPanel } = ToolsPanel;

Then the other change is that I removed <Toolbar> from the CustomToolbar and I added as follows with the enableFilter:
toolbar={ <Toolbar enableFilter={true}><CustomToolbar....../></Toolbar> }

The location of the filter button is not perfect, it appears right next to the drop grouping toolbar.
In addition i was not getting the error that you are getting.
